### ISSUE TITLE
FSDP2 post-backward reduce-scatter hangs on GDN hybrid models (Olmo-Hybrid-7B), 4/4 runs

### ISSUE THREAD

**Issue by abhishekraok** [allenai/OLMo-core#829]
## Summary

SFT of `allenai/Olmo-Hybrid-7B` (GDN + attention hybrid) on a single 8xH100 node
reliably goes silent mid-training. The stalled operation is the FSDP2 post-backward
gradient reduce-scatter. All 8 ranks enqueue the same collective and none of them
completes, so the job holds its GPUs until the process-group timeout fires.

Reproduced 4/4 times across 4 different nodes, at unpredictable steps.

## Symptom

Training logs stop mid-run with no traceback and no NCCL error. With OLMo-core's
default arrangement the process group is built with a long timeout, so nothing aborts
the job — one occurrence held 8 GPUs for 2h40m before a human noticed. Setting the
timeout to 30 minutes produces this:

```
[Rank 4] Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=27732, OpType=_REDUCE_SCATTER_BASE,
         NumelIn=215585536, NumelOut=26948192, Timeout(ms)=1800000)
         ran for 1800003 milliseconds before timing out.

[Rank 4] failure detected by watchdog at work sequence id: 27732
         PG status: last enqueued work: 27753, last completed work: 27731
```

Every rank reports the same `SeqNum` with identical shapes, and every rank has enqueued
past it (`last enqueued 27753` vs `last completed 27731`). So this is not rank divergence
or a shape mismatch: the collective is posted everywhere and never executes. The CPU runs
~20 collectives ahead while the device is stalled, which is why the last logged step is a
few steps beyond where execution actually froze.

With `TORCH_NCCL_TRACE_BUFFER_SIZE` set, the stack is identical on all ranks:

```
#0 reduce_scatter_tensor   torch/distributed/distributed_c10d.py:4591
#1 wrapper                 torch/distributed/c10d_logger.py:83
#2 __call__                torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:125
#3 foreach_reduce          torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:543
#4 decorate_context        torch/utils/_contextlib.py:124
#5 post_backward           torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:566
#6 backward                torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:900
#7 apply                   torch/autograd/function.py:317
```

## Reproductions

| run | hung at step | node | checkpointing | fla |
|---|---|---|---|---|
| 1 | 580 | jupiter-cs-aus-137 | async | 0.5.0 |
| 2 | 350 | jupiter-cs-aus-123 | sync | 0.5.0 |
| 3 | 65  | jupiter-cs-aus-149 | sync | 0.5.0 |
| 4 | 205 | jupiter-cs-aus-219 | async | 0.5.2 |

## What we ruled out

- **Checkpointing.** Identical with `save_async` true and false, and run 3 hung at step 65
  with the first checkpoint scheduled at 172, i.e. before any save had occurred.
- **A bad node.** Four distinct nodes.
- **fla version.** Same signature on 0.5.0 and 0.5.2 (the latter being the version
  OLMo-core mainline moved to in #798).
- **Checkpoint deletion / storage.** Runs used `keep_last_n_checkpoints -1`, so no deletion.

Not ruled out, and the obvious remaining suspects: `torch.compile`, the GDN `fla` kernels
themselves, and their interaction with selected-modules activation checkpointing. We could
not test with compile disabled because the model does not fit at seq 32768 without it
(OOM at ~67 GiB allocated with 12.25 GiB more requested, even after eliminating
fragmentation with `expandable_segments`).

## Configuration

- `allenai/Olmo-Hybrid-7B`, 1 node x 8 H100, seq len 32768
- global batch 1,048,576 tokens (per-device 1, grad accum 4)
- FSDP2, `torch.compile` enabled, activation checkpointing over selected modules
  (`blocks.*.attention` etc. with `determinism_check="none"`), `flash_2`
- OLMo-core pinned at `fa6c5014c9f6e9ee789da2d9c20d5126fee8df0d` (2026-06-22)
- torch 2.10, driven from allenai/open-instruct's `olmo_core_finetune.py`

Happy to re-run with extra instrumentation if that would help — the reproduction is
cheap to trigger (it has never survived past step 580) and we have FlightRecorder wired
up already.

**Comment by abhishekraok** [comment]
## Update: does not reproduce on Blackwell

Ran the identical configuration on a B300 node (`holmes-cs-aus-523`) instead of H100.
It **completed all 861 steps, exit 0**, in ~2h03m.

| platform | node | outcome |
|---|---|---|
| H100 | jupiter-cs-aus-137 | hang at step 580 |
| H100 | jupiter-cs-aus-123 | hang at step 350 |
| H100 | jupiter-cs-aus-149 | hang at step 65 |
| H100 | jupiter-cs-aus-219 | hang at step 205 |
| **B300** | holmes-cs-aus-523 | **861/861, exit 0** |

Same model, same data, same seeds, same open-instruct commit, same hyperparameters. The
Blackwell run cleared every step at which a Hopper run had died, including 580, and was
also considerably faster: 16,392 tok/s/device against 9,073 on H100.

## Why this may be the GDN kernels on Hopper specifically

open-instruct's `pyproject.toml` already carries this note:

> Provides the tilelang backend for fla's gated `chunk_bwd_dqkwg`, which fla routes to on
> **Hopper** with Triton>=3.4.0 (the Triton kernel produces incorrect gradients there; see
> fla-org/flash-linear-attention#640).

That is a gated DeltaNet **backward** kernel, on **Hopper**, already known to misbehave
badly enough to need a workaround — and this hang is in the backward pass, in the
post-backward gradient reduce-scatter. A GDN backward kernel that stalls would block that
rank's stream, so its NCCL kernel never launches and every rank waits in the collective
forever, which matches the observed signature exactly (all ranks enqueued the same work
id, none completed).

**Is gated DeltaNet expected to work on Hopper at all right now, or is it effectively
Blackwell-only?** If the latter, it would be worth stating somewhere: today the failure
is silent and expensive. With the default long process-group timeout, a stalled run holds
its GPUs until someone notices — ours held 8 H100s for 2h40m before we did.

## Caveats

- One Blackwell run against four Hopper failures. The Hopper failure step varies a lot
  (65 to 580), so it is stochastic; a single clean run is strong but not conclusive.
  That said, no Hopper run has ever passed 580.
- **Confound:** the B300 job necessarily used the CUDA 13 image variant, while all the
  Hopper jobs used CUDA 12. So toolkit/torch build version is confounded with GPU
  architecture, and we have not separated the two. If someone can run CUDA 13 on H100,
  that would isolate it.
- Already ruled out on Hopper: checkpointing (hangs identically with `save_async` true and
  false, and the step-65 case preceded any save), single-node hardware (four nodes), and
  the fla version (0.5.0 and 0.5.2 both hang). `torch.compile` could not be tested —
  the model does not fit at seq 32768 without it.

**Comment by abhishekraok** [comment]
## The CUDA-version confound cannot be separated on this fleet

I flagged above that the B300 run necessarily used the CUDA 13 image while every H100 run
used CUDA 12, so toolkit version was confounded with GPU architecture. I tried to separate
them in both directions. Neither is runnable:

| | CUDA 12 | CUDA 13 |
|---|---|---|
| **H100** (ai2/jupiter) | hangs, 4/4 | **not runnable** |
| **B300** (ai2/holmes) | **not runnable** | completes 861/861 |

**CUDA 13 on H100** fails at startup — the driver on jupiter is older than the build needs:

```
CUDA initialization: The NVIDIA driver on your system is too old (found version 12080).
```

**CUDA 12 on B300** fails during `torch.compile` — CUDA 12's nvrtc cannot target the
architecture:

```
nvrtc: error: invalid value for --gpu-architecture (-arch)
```

So only the diagonal is available, and no A/B test on these clusters can isolate one
factor. Anyone picking this up should not spend time trying; the discrimination has to
come from reading the code rather than from running the matrix.

That makes the fla lead more useful than it looked, since it is testable by inspection
rather than by hardware: `chunk_bwd_dqkwg` is routed differently on Hopper with
Triton>=3.4.0, it is a **backward** kernel, this hang is in the backward pass, and the
Triton path there is already known to produce incorrect gradients
(fla-org/flash-linear-attention#640). If someone can confirm whether that routing is
reached in this configuration, that would settle it without needing the missing cells.

Practically, for anyone hitting this today: the two configurations that actually exist are
H100+CUDA12 (hangs) and B300+CUDA13 (works), so running GDN hybrids on Blackwell is the
available workaround whichever factor turns out to be causal.
